File: Airlink/Airlink/Services/HttpsEndpoint.cs
using Airlink.Models.ProfileModel;
using System;
using System.Collections.Generic;
using System.Text;
using Airlink.ViewModels.ProfileViewModel;
using Xamarin.Essentials;
using Xamarin.Forms.Xaml;
using System.Diagnostics;

namespace Airlink.Services
{
    /*
     * Special class to get Tenant Token and Endpoint API for communication
     */
    [XamlCompilation(XamlCompilationOptions.Compile)]

    public class HttpsEndpoint
    {
        public static string ApiEndPoint(string urlType, string deviceName)
        {
            System.Threading.Tasks.Task<string> urlGetTask = SecureStorage.GetAsync("airlinkServer_url");
            System.Threading.Tasks.Task<string> tokenGetTask = null;
            string urlPreModifier;
            string urlPostModifier;
            string deviceTokenKey = "D_" + deviceName;
            switch (urlType)
            {
                case "attributes": //Attributes are not telemetry! They're telemetry without time-series history so 'configuration'
                    urlPreModifier = "/api/v1/"; 
                    tokenGetTask = SecureStorage.GetAsync("deviceAccessToken");
                    urlPostModifier = "/attributes/";
                    break;
                case "advtPost": //Advertising should be posted whether gateway owns device or not, to enable lost device recovery
                    urlPreModifier = "/api/v1/";
                    tokenGetTask = SecureStorage.GetAsync("gatewayAccessToken");
                    urlPostModifier = "/telemetry/";
                    break;
                case "telemetry": //Telemetry only posted for gateway itself or the devices owned by the gateway, individually
                    urlPreModifier = "/api/v1/"; 
                    tokenGetTask = SecureStorage.GetAsync("deviceAccessToken");
                    urlPostModifier = "/telemetry/";
                    break;
                case "provisionDevice": //Provisioning could be done by an agent of the company, so this function may be masked for clients
                    urlPreModifier = "/api/device?accessToken=";
                    tokenGetTask = SecureStorage.GetAsync("deviceAccessToken");
                    urlPostModifier = null;
                    break;
                case "provisionGateway": //Provisioning could be done by an agent of the company, so this function may be masked for clients
                    urlPreModifier = "/api/device?accessToken=";
                    tokenGetTask = SecureStorage.GetAsync("gatewayAccessToken");
                    urlPostModifier = null;
                    break;
                case "getAttributes":
                    urlPreModifier = "/api/v1/";
                    tokenGetTask = SecureStorage.GetAsync(deviceTokenKey);
                    urlPostModifier = "/attributes";
                    break;
                case "serverScope":
                    urlPreModifier = "/api/plugins/telemetry/DEVICE/";
                    tokenGetTask = SecureStorage.GetAsync("deviceUUID");
                    urlPostModifier = "/SERVER_SCOPE";
                    break;
                default:
                    urlPreModifier = "";
                    urlPostModifier = "";
                    break;
            }
            string token = (tokenGetTask == null) ? "" : tokenGetTask.Result;
            string _endPointApi = urlGetTask.Result + urlPreModifier + token + urlPostModifier;
            return _endPointApi;
        }
    }
}

File: airlink_flutter/lib/services/HttpsEndpoint.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

class HttpsEndpoint {
  Future<String> apiEndpoint(String action) async {
    const storage = FlutterSecureStorage();

    final serverUrl = await storage.read(key: 'serverUrl');

    String urlPreModifier;
    String token;
    String urlPostModifier;

    switch (action) {
      case 'login':
        urlPreModifier = '/api';
        token = '';
        urlPostModifier = '/auth/login';
        break;
      case 'provisionGateway':
        urlPreModifier = '/api/device?accessToken=';
        final gatewayAccessToken =
            await storage.read(key: 'gatewayAccessToken');
        token = gatewayAccessToken.toString();
        urlPostModifier = '';
        break;
      case 'provisionDevice':
        urlPreModifier = '/api/device?accessToken=';
        final deviceAccessToken = await storage.read(key: 'deviceAccessToken');
        token = deviceAccessToken.toString();
        urlPostModifier = '';
        break;
      case 'serverScope':
        urlPreModifier = '/api/plugins/telemetry/DEVICE/';
        final deviceUUID = await storage.read(key: 'deviceUUID');
        token = deviceUUID.toString();
        urlPostModifier = '/SERVER_SCOPE';
        break;
      case 'getAttributes':
        urlPreModifier = '/api/v1/';
        final deviceAccessToken = await storage.read(key: 'deviceAccessToken');
        token = deviceAccessToken.toString();
        urlPostModifier = '/attributes';
        break;
      case 'telemetry':
        urlPreModifier = '/api/plugins/telemetry/';
        token = '';
        urlPostModifier = '';
        break;
      case 'advtPost':
        urlPreModifier = '/api/v1/';
        final gatewayAccessToken =
            await storage.read(key: 'gatewayAccessToken');
        token = gatewayAccessToken.toString();
        urlPostModifier = '/telemetry';
        break;
      case 'getAttributesByScope':
        urlPreModifier = '/api/plugins/telemetry/';
        token = '';
        urlPostModifier = '';
        break;
      case 'saveEntityAttributes':
        urlPreModifier = '/api/plugins/telemetry/';
        token = '';
        urlPostModifier = '';
        break;
      case 'getTenantDevice':
        urlPreModifier = '/api/tenant/devices?deviceName=';
        urlPostModifier = '';
        token = '';
        break;
      default:
        urlPreModifier = '';
        token = '';
        urlPostModifier = '';
        break;
    }

    String _endPointApi =
        '${serverUrl}${urlPreModifier}${token}${urlPostModifier}';
    return _endPointApi;
  }
}
